---
layout: default
categories:
 - app
---

<div class="row">
	<div class="col-md-3">
		<div class="spacer"></div>
		<div class="io-container title">
			Phylogeny
		</div>
		<div class="spacer"></div>
		<div class="row">
			<div class="col-xs-6 io-container">
				<div>
					<svg id="date-input"></svg>
				</div>
				<div>
					<h5 id="legend-title">Legend</h5>
					<svg id="legend"></svg>
				</div>
				<div class="spacer"></div>
			</div>
			<div class="col-xs-6 io-container">
				<form class="form-horizontal form-small">
					<div class="form-group">
							<label class="col-xs-4 control-label">Color by</label>
							<div class="col-xs-8 controls">
								<select id="coloring" class="form-control">
                  {% for color_option in site.data.color_options %}
                    {% if page.coloring contains color_option[0] %}
                    <option value={{color_option[0] }}>{{ color_option[1] }}</option>
                    {% endif %}
                  {% endfor %}
								</select>
							</div>
					</div>
					<div class="form-group">
						<label class="col-xs-4 control-label">Or</label>
						<div class="col-xs-8 controls">
							<input id="gt-color" type="search" placeholder= "{{ page.gtplaceholder }}" class="form-control">
						</div>
					</div>
					{% if page.coloring contains 'HI' %}
					<div class="form-group" id="HIcontrols" style="display: none">
						<label class="col-xs-4 control-label">HI color:</label><br>
						<div class="col-xs-8 controls">
							<form action="" id="HImodel">
								<input type="radio" name="HImodel" id="HImodel_measured"  checked="checked"  value="measured"> measurements<br>
 								<!-- <input type="radio" name="HImodel" id="HImodel_tree" value="tree"> tree model<br> --> <!-- off in favor of subs model -->
 								<input type="radio" name="HImodel" id="HImodel_mutation" value="mutation"> substit. model
							</form>
						</div>
						<label class="col-xs-4 control-label">Corrections:</label><br>
						<div class="col-xs-8 controls">
							<div class="bigspacer"></div>
							<input type="checkbox" id="serum" value="serum" checked="checked"> serum potency<br>
							<input type="checkbox" id="virus" value="virus"> virus avidity
						</div>
						<label class="col-xs-4 control-label">Sera:</label><br>
						<div class="col-xs-8 controls">
							<div class="bigspacer"></div>
							<div id="sera"></div>
						</div>
					</div>
					{% endif %}
				</form>

				{% for res in site.data.restrictions %}
				<form class="form-horizontal form-small">
					<div class="form-group">
						<label class="col-xs-4 control-label">{{res.label}}</label>
						<div class="col-xs-8 controls">
							<select id="{{res.id}}" class="form-control">
								{% for reg in res.fields %}
								<option value="{{ reg[0] }}">{{reg[1] }}</option>
								{% endfor %}
							</select>
						</div>
					</div>
				</form>
				{% endfor %}

        <form class="form-horizontal form-small">
          <div class="form-group">
            <label class="col-xs-4 control-label">Search</label>
            <div class="col-xs-8 controls">
              <div id="straininput"></div>
              <div class="smallspacer"></div>
              <button id="searchinputclear" type="button" class="btn btn-default">clear search</button>
            </div>
          </div>
        </form>

        <form class="form-horizontal form-small">
          <div class="form-group">
            <label class="col-xs-4 control-label">Options</label>
            <div class="col-xs-8 controls">
              <div class="checkbox">
                <label>
                  {% if page.site == 'WHO' %}
                  <input type="checkbox" id="branchlabels" checked="checked"> branch labels
                  {% else %}
                  <input type="checkbox" id="branchlabels"> branch labels
                  {% endif %}
                </label>
              </div>
              <div class="checkbox">
                <label>
                  <input type="checkbox" id="timetree" checked="checked"> timetree
                </label>
              </div>
            </div>
          </div>
        </form>

				<!-- <textarea id="seqinput" class="form-control" placeholder="search sequences... (plain or fasta)"></textarea>
				<div class="spacer"></div> -->

        <form class="form-horizontal form-small">
          <div class="form-group">
            <div class="col-xs-2"></div>
            <div class="col-xs-8 controls">
              <button id="reset" type="button" class="btn btn-default" style="padding:8px 30px 8px 30px;">reset layout</button>
            </div>
            <div class="col-xs-2"></div>
          </div>
        </form>

			</div>
		</div>
	</div>
	<div class="col-md-9">
		<div class="treeplot-container" id="treeplot-container">
			<svg id="treeplot"></svg>
		</div>
	</div>

</div>

<div class="row">
	<div class="col-md-12">
		<div class="bigspacer"></div>
	</div>
</div>

<div class="row">
	<div class="col-md-3">
		<div class="io-container title">
			Frequencies
		</div>
		<div class="spacer"></div>
		<div class="row">
			<div class="col-xs-6 io-container">
				<form class="form-horizontal form-small">
					<div class="form-group">
						<div class="col-xs-12">
							<input id="gtspec" type="search" class="form-control" value="{{ page.freqdefault }}">
						</div>
					</div>
				</form>
			</div>
			<div class="col-xs-6 io-container">
				<div>
					<button id="plotfreq" type=button class="btn btn-default">plot frequencies</button>
					<div class="spacer"></div>
					<button id="downloadfreq" type=button class="btn btn-default">download</button>
				</div>
			</div>
		</div>
		<div class="spacer"></div>
		Input mutations as position+amino acid, i.e. <span class="highlight">159Y</span>, clades as clade name,
		i.e. <span class="highlight">3c2.a</span> and add locations as <span class="highlight">/AS</span>,
		<span class="highlight">/NA</span>, <span class="highlight">/EU</span>, <span class="highlight">/OC</span>,
		i.e. <span class="highlight">159Y/AS</span>. By default, positions are interpreted as residues in HA1. You can specify the subunit as <span class="highlight">HA2:18V</span>. Alternatively, simply click on variable positions in the graph below.
		<div class="spacer"></div>
	</div>
	<div class="col-md-9">
		<div class="freqplot-container">
			<div id="gtchart"></div>
		</div>
		<div class="entropy-container">
			<div id="entropy"></div>
		</div>
	</div>
</div>

<div class="row">
	<div class="col-md-12">
		<div class="bigspacer"></div>
	</div>
</div>

<div class="row">
	<div class="col-md-3">
		<div class="io-container title">
			Feature explanation
		</div>
	</div>
	<div class="col-md-9" align="justify">
		 <a href="/help/#frequencies">Click here</a> for help with the nextflu interface.<p>
		<h4>HI data</h4>
		<p>HI data can be displayed as color on the tree or viewed via the tool tips that
		show when moving the mouse over a circle corresponding to a virus. To explore the HI titer data,
		select <span class="highlight">HI distance from focus</span> in the
		<span class="highlight">color by</span> menu and click on one of the available
		reference viruses indicated by grey squares. The tree will then be colored
		by log2 distance from this reference virus. The coloring either reflects the
		the direct measurements of HI titers provided by the WHO collaborating centers
		(notably the annual and interim reports by the
		<a href="http://www.crick.ac.uk/research/worldwide-influenza-centre/annual-and-interim-reports/">Crick Worldwide Influenza Centre</a> in London), or models that are fit to these data. Whether the raw data, the tree model
		and the mutation model are used to color the tree can be chosen via the radio button on the left.
		If more than one measurement is available, we take the average over all available measurements.
		In the process of fitting the models, column (serum potency) and row
		(virus avidities) effects are estimated. These corrections can be
		subtracted from the raw measurements to remove noise.
		To see all measurements of a virus relative to the chosen reference
		virus, put the mouse over that virus and a info box (tooltip) will pop
		up with a table that lists all measurements (and the autologous titers
		for the sera to facilitate interpretation) and the model predictions.</p>
		<p>The tree can also be colored by cumulative antigenic change
		-- similar to dimension 1 in antigenic cartography.</p>
		<h4>Phylogenetic tree</h4>
		<p> Use the <a href="/help/#date_slider"><b>date slider</b></a> to select viruses sampled within the time interval indicated. The size of the interval can be changed by grabbing the left end of the bar with the mouse, to move the interval, use the right end of the slider.</p>
		<p> Use the <a href="/help/#epitope_mutations"><b>drop down menu</b></a> to color viruses by number of <span class="highlight"><a href="/help/#epitope_mutations">epitope mutations</a></span>,
		<span class="highlight"><a href="/help/#nonepitope_mutations">non-epitope mutations</a></span> or
		<span class="highlight"><a href="/help/#rbs">receptor binding mutations</a></span>
		relative to root, or to color viruses by <span class="highlight"><a href="/help/#lbi">local branching index</a></span> or
		<span class="highlight"><a href="/help/#geo">geographic region</a></span>.
		<p> Use the <a href="/help/#genotype"><b>input box</b></a> to specify positions to color viruses by genotype. Amino acid positions must be
		separated by a comma (e.g. <span class="highlight">159,225</span>). The default is <span class="highlight">HA1</span>, to color by amino acid sequence in other regions use <span class="highlight">HA2:18</span> or <span class="highlight">SigPep:6</span>. To color by nucleotide sequence, use <span class="highlight">nuc:527</span>.
		<p> <a href="/help/#mouse_tip"><b>Mouse over</b></a> a tip to show virus name, location and features.
		<p> <a href="/help/#mouse_branch"><b>Mouse over</b></a> a branch to graph the frequency of the correponding clade trajectory below or <b>click</b>
		on a branch to zoom into its descendent clade. The tool tip will show amino acid mutations on this branch.
		<p> To <b>restrict</b> the displayed viruses to <b>certain geographic regions</b>, select the region in the drop down menu labeled <span class="highlight">region</span>.
		<h4>Frequencies</h4>
		Enter a mutation or genotype above (e.g. <span class="highlight">225D</span>) and click
		<b>plot frequencies</b> to show estimated frequency of this mutation through time. In addition, geographic regions can be specified by adding AS
		(Asia), NA (North America), EU (Europe), or OC (Oceania) as <span class="highlight">159S/225D/AS</span>.
		Several genotypes can be entered simultaneously when separated by commas (e.g.
		<span class="highlight">225D, 159S/225D/AS</span> will graph the global frequency of 225D and the frequency
		of strains containing both 159S and 225D in Asia). Instead of a genotype, the common clades 3c3, 3c3.a, 3c2, 3c2.a can be used. Positions with very little variation are omitted. Beware
		that region specific frequencies are noisy.
		<h4>Variability</h4>
		The second plot shows the variation in the multiple sequence alignment used to construct the tree. High bars indicated variable positions. Clicking on those bars will color the tree by amino acid at this position and plot the frequencies of the corresponding amino acids.
	</div>
</div>

<div class="row">
	<div class="col-md-12">
		<div class="bigspacer"></div>
	</div>
</div>

<div class="row">
	<div class="col-md-3">
		<div class="io-container title">
			Rationale and details
		</div>
		<div class="spacer"></div>
	</div>
	<div class="col-md-9" align="justify">
		<p> <b>Epitope mutations</b> are based on
		<a href="http://www.nature.com/nature/journal/v289/n5796/abs/289366a0.html">HA structure and exposed residues</a>.
		Multiple recent mutations at epitope
		sites have been <a href="http://www.nature.com/nature/journal/v507/n7490/abs/nature13087.html">
		suggested</a> to be predictive for strains dominating future seasons. Similarly, mutations
		outside of these epitopes -- termed <b>non-epitope sites</b> --- tend to be damaging and are suggested to
		be predictive of clade contraction.</p>
		<p>Antigenic evolution has been <a href="http://www.sciencemag.org/content/342/6161/976.abs">shown</a>
		to depend primarily on substitutions surrounding the receptor binding site of HA1.
		These seven positions (145, 155, 156, 158, 159, 189, 193 in HA1 numbering) are referred to here
		as <b>receptor binding positions</b> and changes at these positions could correspond to
		large changes in antigenic properties.</p>
		<p>The <b>local branching index</b> is the exponentially weighted tree length surrounding
		a node, which is associated with <a href="http://elifesciences.org/content/3/e03568">
		rapid branching and expansion of clades</a>. A more detailed explanation is available
		<a href="https://neherlab.wordpress.com/2014/11/11/now-out-in-elife-predicting-evolution/">here</a>.
		Retrospective analysis has shown that LBI correlates with clade growth.
		</p><p>
		<b>Frequencies</b> are estimated as maximum likelihood trajectories that penalize rapid changes in
		frequency and slope. The frequencies of large clades or abundant genotypes have sufficiently
		many observations to by robust, while frequencies of rare mutations can't be reliably
		estimated.
	</div>
</div>

<div class="row">
	<div class="col-md-12">
		<div class="bigspacer"></div>
	</div>
</div>


<div class="row">
	<div class="col-md-3"></div>
	<div class="col-md-9">
		<hr>

    <p>
    Built and maintained by <a href="https://neherlab.org/">Richard Neher</a>,
    <a href="http://bedford.io/team/trevor-bedford/">Trevor Bedford</a> and
    <a href="http://bedford.io/team/barney-potter/">Barney Potter</a>.
    All <a href="http://github.com/blab/nextflu" target="_blank">source
    code</a> is freely available under the terms of the <a
    href="http://github.com/blab/nextflu/blob/master/LICENSE.txt" target="_blank">GNU Affero General
    Public License</a>. Data <span id="updated"></span><span id="commit"></span>.
    </p>
    <p>
    This work is made possible by the <a href="http://gisaid.org" target="_blank">GISAID Initiative</a>
    and the open sharing of genetic data by influenza research groups from all over the world. <a
    href="/acknowledgements/">We gratefully acknowledge their contributions.</a> Titer data
    was generated by the <a
    href="https://www.cdc.gov/flu/">Influenza Division at the US Centers for Disease Control and
    Prevention</a>, the <a href="http://www.crick.ac.uk/research/worldwide-influenza-centre">Worldwide
    Influenza Centre at the Francis Crick Institute</a>, the <a href="http://www.vidrl.org.au/">Victorian Infectious Diseases Reference
    Laboratory at the Australian Peter Doherty Institute for Infection and Immunity</a> and the
    <a href="https://www.niid.go.jp/niid/en/flu-e.html">Influenza Virus Research Center at the Japan
    National Institute of Infectious Diseases</a>.
    </p>
    <p>
    Special thanks to Charlton Callender, John Huddleston,
    Jackie Katz, Dave Wentworth, Becky Garten, Vivien Dugan, Xiyan Xu, Elizabeth Neuhaus, Sujatha Seenu,
    John McCauley, Rod Daniels, Vicki Gregory,
    Kanta Subbarao, Ian Barr, Aeron Hurt,
    Takato Odagiri, Shinji Watanabe, Tomoko Kuwahara,
    Michael Lässig, Marta &Lstrok;uksza, Richard Reeve,
    Colin Russell, Sebastian Maurer-Stroh, Peter Bogner, David Lehman
    and the <a href="https://www.aerobatic.com/">Aerobatic</a> team for help on many aspects of nextflu.
    </p>
    <p>
    Please cite: <a href="https://doi.org/10.1093/bioinformatics/btv381">Neher RA, Bedford T. 2015.
    nextflu: real-time tracking of seasonal influenza virus evolution in humans. Bioinformatics
    10.1093/bioinformatics/btv381</a>.
    </p>

	</div>
</div>

<div class="row">
	<div class="col-md-3"></div>
	<div class="col-md-9">
		<hr>
		<div class="center text-gray smallnote">
			&copy; 2015-2018 Trevor Bedford and Richard Neher
		</div>
	</div>
</div>

<div class="row">
	<div class="col-md-12">
		<div class="bigspacer"></div>
	</div>
</div>

<div class="row">
	<div class="col-md-3"></div>
	<div class="col-md-9">
		<div class="center">
			<a href="http://www.fredhutch.org/"><img style="display:inline;" width=80 src="/images/fred-hutch-logo-small.png"></a>
			&nbsp; &nbsp; &nbsp;
			<a href="http://www.eb.tuebingen.mpg.de/"><img style="display:inline;" width=60 src="/images/max-planck-logo-small.png"></a>
			&nbsp; &nbsp; &nbsp;
			<a href="https://www.nih.gov/"><img style="display:inline;" width=52 src="/images/nih-logo-small.png"></a>
			&nbsp; &nbsp; &nbsp;
			<a href="https://erc.europa.eu/"><img style="display:inline;" width=60 src="/images/erc-logo-small.png"></a>
			&nbsp; &nbsp; &nbsp;
      <a href="http://www.openscienceprize.org/"><img style="display:inline;" width=90 src="/images/osp-logo-small.gif"></a>
		</div>
	</div>
</div>

<div class="row">
	<div class="col-md-12">
		<div class="bigspacer"></div>
	</div>
</div>

{{ content }}

<script src="/js/auspice.js"></script>
