---
layout: default
title: nextflu
---

<div class="row">
	<div class="col-md-3">
		<div class="logo-container">
			<div class="bigtitle">
				<span class="c2">n</span><span class="c3">e</span><span class="c4">x</span><span class="c5">t</span><span class="c6">f</span><span class="c7">l</span><span class="c8">u</span>
			</div>
			<div class="subtitle-box">
				Real-time tracking of seasonal influenza virus evolution in humans
			</div>
			<div class="spacer"></div>		
		</div>
		<div class="io-container">

			<form class="form-inline form-small">
				<div class="form-group">
					<label class="control-label">Virus</label>
					<select id="virus_select" class="form-control" onChange="document.location = this.value" value="GO">
						<option value="/H3N2/" 
							{% if page.virus == "H3N2" %}selected="selected"{% endif %}
							>A/H3N2</option>
						<option value="/H1N1pdm/"
							{% if page.virus == "H1N1pdm" %}selected="selected"{% endif %}
							>A/H1N1pdm</option>
						<option value="/Vic/"
							{% if page.virus == "Vic" %}selected="selected"{% endif %}
							>B/Vic</option>
						<option value="/Yam/"
							{% if page.virus == "Yam" %}selected="selected"{% endif %}
							>B/Yam</option>
					</select>
				</div>
			</form>		
					
		</div>
		<div class="bigspacer"></div>	
		<div class="io-container title">
			HA phylogeny
		</div>
		<div class="spacer"></div>		
		<div class="row">
			<div class="col-xs-6 io-container">
				<div>
					<svg id="date-input"></svg>
				</div>
				<div>
					<h5 id="legend-title">Legend</h5>
					<svg id="legend"></svg>
				</div>
				<div class="spacer"></div>
			</div>
			<div class="col-xs-6 io-container">			
				<form class="form-horizontal form-small">
					<div class="form-group">
							<label class="col-xs-4 control-label">Color by</label>
							<div class="col-xs-8 controls">
								<select id="coloring" class="form-control">
									{% if page.coloring contains 'ep' %}
									<option value="ep">epitope mutations</option>
									{% endif %}
									{% if page.coloring contains 'ne' %}
									<option value="ne">non-epitope mutations</option>
									{% endif %}
									{% if page.coloring contains 'rb' %}
									<option value="rb">receptor binding mutations</option>
									{% endif %}
									{% if page.coloring contains 'lbi' %}
									<option value="lbi">local branching index</option>
									{% endif %}
									{% if page.coloring contains 'region' %}
									<option value="region">geographic region</option>
									{% endif %}
									{% if page.coloring contains 'dfreq' %}
									<option value="dfreq">frequency change</option>
									{% endif %}
									{% if page.coloring contains 'HI %}
									<option value="cHI">antigenic advance</option>
									{% endif %}
								</select>
							</div>
					</div>
					<div class="form-group">
						<label class="col-xs-4 control-label">Or</label>
						<div class="col-xs-8 controls">
							<input id="gt-color" type="search" placeholder= "{{ page.gtplaceholder }}"  class="form-control">
						</div>
					</div>
					{% if page.coloring contains 'HI %}
					<div class="form-group">
						<label class="col-xs-4 control-label"></label>
						<div class="col-xs-8 controls">
						<input type="checkbox" id="HIprediction" value="pred"> use predicted HI<br>
						<input type="checkbox" id="serum" value="serum"> correct for potency<br>
						<input type="checkbox" id="virus" value="virus"> correct for virus effect<br>
						</div>
					</div>		
					{% endif %}	
				</form>
				<div class="spacer"></div>
				
				<form class="form-horizontal form-small">
					<div class="form-group">
						<label class="col-xs-4 control-label">Region</label>
						<div class="col-xs-8 controls">
							<select id="region" class="form-control">
								<option value="all">all</option>
								<option value="Africa">Africa</option>
								<option value="China">China</option>
								<option value="Europe">Europe</option>
								<option value="JapanKorea">Japan / Korea</option>
								<option value="NorthAmerica">North America</option>
								<option value="Oceania">Oceania</option>
								<option value="SouthAmerica">South America</option>
								<option value="SouthAsia">South Asia</option>								
								<option value="SoutheastAsia">Southeast Asia</option>
								<option value="WestAsia">West Asia</option>
							</select>
						</div>
					</div>
				</form>
				<div id="search"></div>
				<div class="spacer"></div>
				<div>
					<button id="reset" class="btn btn-default">reset layout</button>
				</div>
				<div class="spacer"></div>
			</div>					
		</div>
	</div>
	<div class="col-md-9">
		<div class="bigspacer"></div>
		<div class="treeplot-container">
			<svg id="treeplot"></svg>
		</div>
	</div>
</div>

<div class="row">
	<div class="col-md-12">
		<div class="bigspacer"></div>
		<div class="bigspacer"></div>
	</div>
</div>

<div class="row">
	<div class="col-md-3">
		<div class="io-container title">
			Frequencies
		</div>
		<div class="spacer"></div>
		<div class="row">			
			<div class="col-xs-6 io-container">	
				<form class="form-horizontal form-small">
					<div class="form-group">
						<div class="col-xs-12">
							<input id="gtspec" type="search" class="form-control" value="{{ page.freqdefault }}">
						</div>
					</div>
				</form>	
			</div>
			<div class="col-xs-6 io-container">					
				<div>
					<button id="plotfreq" type=button class="btn btn-default">plot frequencies</button>
				</div>
			</div>
		</div>
		<div class="spacer"></div>
		Input mutations as position+amino acid, i.e. <span class="highlight">159Y</span>, clades as clade name, 
		i.e. <span class="highlight">3c2.a</span> and add locations as <span class="highlight">/AS</span>, 
		<span class="highlight">/NA</span>, <span class="highlight">/EU</span>, <span class="highlight">/OC</span>, 
		i.e. <span class="highlight">159Y/AS</span>
	</div>
	<div class="col-md-9">
		<div class="freqplot-container">		
			<div id="gtchart"></div>
		</div>
	</div>	
</div>

<div class="row">
	<div class="col-md-12">
		<div class="bigspacer"></div>
	</div>
</div>

<div class="row">
	<div class="col-md-3">
		<div class="io-container title">
			Feature explanation
		</div>	
	</div>
	<div class="col-md-9">
		<h4>Phylogenetic tree</h4>
		<p> Use the <a href="/help/#date_slider"><b>date slider</b></a> to highlight viruses sampled within 12 month of the date displayed.
		<p> Use the <a href="/help/#epitope_mutations"><b>drop down menu</b></a> to color viruses by number of <span class="highlight"><a href="/help/#epitope_mutations">epitope mutations</a></span>, 
		<span class="highlight"><a href="/help/#nonepitope_mutations">non-epitope mutations</a></span> or 
		<span class="highlight"><a href="/help/#rbs">receptor binding mutations</a></span>
		relative to root, or to color viruses by <span class="highlight"><a href="/help/#lbi">local branching index</a></span> or 
		<span class="highlight"><a href="/help/#geo">geographic region</a></span>.
		<p> Use the <a href="/help/#genotype"><b>input box</b></a> to specify positions to color viruses by genotype. Amino acid positions must be 
		separated by a comma (e.g. <span class="highlight">159,225</span>).
		<p> <a href="/help/#mouse_tip"><b>Mouse over</b></a> a tip to show virus name, location and features.
		<p> <a href="/help/#mouse_branch"><b>Mouse over</b></a> a branch to graph the frequency of the correponding clade trajectory below or <b>click</b>
		on a branch to zoom into its descendent clade.
		<h4>Frequencies</h4>
		Enter a mutation or genotype above (e.g. <span class="highlight">225D</span>) and click 
		<b>plot frequencies</b> to show estimated frequency of this mutation through time. Complex genotypes can be specified by combining mutations as 
		<span class="highlight">159Y/225D</span>. In addition, geographic regions can be specified by adding AS 
		(Asia), NA (North America), EU (Europe), or OC (Oceania) as <span class="highlight">159S/225D/AS</span>.
		Several genotypes can be entered simultaneously when separated by commas (e.g. 
		<span class="highlight">225D, 159S/225D/AS</span> will graph the global frequency of 225D and the frequency 
		of strains containing both 159S and 225D in Asia). Instead of a genotype, the common clades 3c3, 3c3.a, 3c2, 3c2.a can be used. Positions with very little variation are omitted. Beware 
		that region specific frequencies are noisy. <a href="/help/#frequencies">Click here for help with the interface.</a>		
	</div>	
</div>

<div class="row">
	<div class="col-md-12">
		<div class="bigspacer"></div>
	</div>
</div>

<div class="row">
	<div class="col-md-3">
		<div class="io-container title">
			Rationale and details
		</div>	
		<div class="spacer"></div>	
	</div>
	<div class="col-md-9">
		<p> <b>Epitope mutations</b> are based on 
		<a href="http://www.nature.com/nature/journal/v289/n5796/abs/289366a0.html">HA structure and exposed residues</a>. 
		Multiple recent mutations at epitope
		sites have been <a href="http://www.nature.com/nature/journal/v507/n7490/abs/nature13087.html">
		suggested</a> to be predictive for strains dominating future seasons. Similarly, mutations
		outside of these epitopes -- termed <b>non-epitope sites</b> --- tend to be damaging and are suggested to 
		be predictive of clade contraction.</p>
		<p>Antigenic evolution has been <a href="http://www.sciencemag.org/content/342/6161/976.abs">shown</a> 
		to depend primarily on substitutions surrounding the receptor binding site of HA1. 
		These seven positions (145, 155, 156, 158, 159, 189, 193 in HA1 numbering) are referred to here
		as <b>receptor binding positions</b> and changes at these positions could correspond to 
		large changes in antigenic properties.</p>
		<p>The <b>local branching index</b> is the exponentially weighted tree length surrounding 
		a node, which is associated with <a href="http://elifesciences.org/content/3/e03568"> 
		rapid branching and expansion of clades</a>. A more detailed explanation is available 
		<a href="https://neherlab.wordpress.com/2014/11/11/now-out-in-elife-predicting-evolution/">here</a>. 
		Retrospective analysis has shown that LBI correlates with clade growth. 
		</p><p>
		<b>Frequencies</b> are estimated as maximum likelihood trajectories that penalize rapid changes in 
		frequency and slope. The frequencies of large clades or abundant genotypes have sufficiently 
		many observations to by robust, while rare mutations or combinations thereof can't be reliably
		estimated.
	</div>
</div>

<div class="row">
	<div class="col-md-3"></div>
	<div class="col-md-9">
		<hr>
		Built with love by <a href="http://bedford.io">Trevor Bedford</a> and 
		<a href="https://neherlab.wordpress.com/">Richard Neher</a>. This work is made possible by the 
		<a href="http://gisaid.org" target="_blank">GISAID Initiative</a> and the open sharing of genetic data by influenza research 
		groups from all over the world. <a href="/acknowledgements/">We gratefully acknowledge their contributions.</a>
		Give us a shout at <a href="https://twitter.com/trvrb" target="_blank">@trvrb</a> or 
		<a href="https://twitter.com/richardneher" target="_blank">@richardneher</a> with questions or comments. All 
		<a href="http://github.com/blab/nextflu" target="_blank">source code</a> is freely available under the terms of the 
		<a href="http://github.com/blab/nextflu/blob/master/LICENSE.txt" target="_blank">GNU Affero General Public License</a>. Data 
		updated <span id="updated"></span> and processed with commit <span id="commit"></span>.		
	</div>
</div>

<div class="row">
	<div class="col-md-3"></div>
	<div class="col-md-9">
		<hr>		
		<div class="center text-gray smallnote">
			&copy; 2015 Trevor Bedford and Richard Neher
		</div>
	</div>
</div>

<div class="row">
	<div class="col-md-12">
		<div class="bigspacer"></div>
	</div>
</div>

<script src="/js/{{ page.virus }}.js"></script>
<script src="/js/auspice.js"></script>
