---
layout: report
title: Seasonal influenza circulation patterns and projections for Feb 2018 to Feb 2019
linktitle: Feb 2018
categories:
 - reports
---

<div class="row">
    <div class="col-md-3">
        <div class="io-container title">
            A/H3N2
        </div>
        <div class="spacer"></div>
    </div>
    <div class="col-md-9">

        <p><b>H3N2 diversity has largely been replaced by subclades A1b, A2 and A3 within 3c2.A.
        Subclades A1b and A2 predominate in the population and each shows increases in frequency, mutations at epitope sites and evidence for minor changes to antigenic phenotype.
        Clade A1b may be marginally fitter than clade A2, but we expect both clades to persist into the future without a clear immediate winner.</b>

        <div class="spacer"></div>

        <p>We base our primary analysis on a set of viruses collected between Mar 2016 and Jan 2018, comprising >1000 viruses per month in Dec 2016 to Mar 2017 and 100-200 viruses per month more recently.
        We use all available data when estimating frequencies of mutations and weight samples appropriately by regional population size and relative sampling intensity to arrive at a putatively unbiased global frequency estimate.
        Phylogenetic analyses are based on a representative sample of about 2000 viruses.

        {% embed_image /reports/figures/feb-2018/h3n2_counts.png 12 %}

        <p>Clades 3c3.A and 3c3.B have hardly been observed for over 1 year and the great majority of H3N2 viruses now fall into clade 3c2.A or its subclades.
        Within 3c2.A, the clade 3c2.A1 (bearing mutation N171K) is now globally at a frequency of about 40\%, while North America is dominated by viruses outside this subclade.

        <p>We observe substantial diversification within the 3c2.A clade over the past year.
        There are 5 major subclades within 3c2.A that have reached appreciable frequency, three deriving from basal clade 3c2.A and two deriving from clade 3c2.A1.
        All of these clades possess mutations at previously characterized epitope sites.
        These clades are characterized by the following mutations and highlighted in the phylogenetic tree:

        <ol>
          <li>Clade 3c2.A1a: 135K, HA2:150E (previously called clade 4)</li>
          <li>Clade 3c2.A1b: 92R, 311Q (previously called clade 5)</li>
          <li>Clade 3c2.A2: 131K, 142K, 261Q (previously called clade 3)</li>
          <li>Clade 3c2.A3: 121K, 144K (previously called clade 2)</li>
          <li>Clade 3c2.A4: 53N, 144R, 171K, 192T, 197H (previously called clade 1)</li>
        </ol>

        Throughout the following, we drop the 3c2 to abbreviate these names to A1a, A1b, A2, A3 and A4.

        {% embed_image /reports/figures/feb-2018/h3n2_tree_clades_with_timetree.png 12 %}

        <p>In the course of 2017, all of these clades had risen in frequency to dominate the H3N2 population.
        However, clades A1a and A4 have recently declined leaving clades A1b, A2 and A3 to make up the majority of circulating viruses.
        Clade A4 was at high frequency in Oceania in the recent in Southern Hemisphere season, but is hardly observed in the current Northern Hemisphere season.
        Clade A1a is close to extinction as well.
        The remaining three clades (A1b, A2 and A3) have distinct geographic distributions.
        Clade A1b is estimated at ~30% of the global population, down from a peak of nearly 60%.
        Clade A2 is estimated at ~40% of the global population, but is dominating the current North American season and is increasingly observed in Europe.
        Clade A3 is estimated at a steady global frequency of about 25%.

        {% embed_image /reports/figures/feb-2018/h3n2_freq_major_clades.png 12 %}

        <p>Within these 5 clades, there have been frequent parallel mutations, particularly at HA1 sites 121, 135 and 142.
        The mutation T135K mutation in clade A3 is responsible for its recent resurgence in frequency.
        Clade A1b shows a similar pattern with mutations T135K and T135N making up two increasing subclades within the A1b clade.

        <p>Recent clade A2 viruses descend from a reassortment event about one year ago in which the HA and PB1 segments of a clade A2 virus moved into the background of a clade A1a virus (hence these viruses show up as HA clade A2).
        The neuraminidase of clade A1a differs from clade A2 by I176M, N329S, D339N, and P386S.
        The original neuraminidase of clade A2 and clade A3 viruses were outcompeted by the reassortant.

        {% embed_image /reports/figures/feb-2018/h3n2_ha_na_reassortment.png 12 %}

        <p>These four recently emerged variants constitute the most successful recent subclades.
        Clade A1b viruses with 135K had increased rapidly in frequency in fall 2017, but are now declining globally.
        However, reassortant clade A2 viruses have completely driven the recent growth of A2 and these viruses are rapidly increasing in multiple regions.

        {% embed_image /reports/figures/feb-2018/h3n2_freq_emerging_clades.png 12 %}

        <p>All 5 of these clades appear somewhat drifted according to antigenic
        analysis (<a href="http://bedford.io/papers/neher-titers/">Neher et al 2016</a>) of HI measurements provided by the Influenza Division at the US Centers for Disease Control and Prevention.
        Antigenic differences from the future Southern Hemisphere vaccine strain A/Singapore/Infimh-16-0019/2016 to circulating viruses are estimated to be between 0.4 and 2.2 antigenic units (or approximately a four-fold HI dilution).

        {% embed_image /reports/figures/feb-2018/h3n2_tree_titers.png 12 %}

        <p>Recent behavior of cocirculating clades A1b, A2 and A3 would suggest that clades A1b and A2 are the most competitive.
        The highest local branching index (LBI) (<a href="https://elifesciences.org/content/3/e03568">Neher et al. 2014</a>) is observed at the base in clade A1b concomitant with the T135K mutation (potential loss of glycosylation, minor antigenic impact) and in clade A2 following reassortment with an evolved NA background.
        With clade A2 HA, there are no recent mutations that suggest adaptation, but the reassortment event has potentially changed intrinsic fitness and antigenic properties of NA.
        Clade A2 has risen extremely rapidly in North America, but it is not yet clear whether similar dynamics play out in Asia.
        However, while there is not large amounts of data, reassortant clade A2 viruses now appear to be spreading in Asia.
        Clade A1b currently appears at higher frequency than clade A2 in Asia.
        However, clade A1b is declining, while clade A2 is increasing.
        The complete dominance of A2 viruses in North America may not strongly reflect future trends, as successful lineages often lead in Asia (<a href="http://bedford.io/papers/bedford-global-circulation/">Bedford et al 2015</a>).
        In a recent example, in spring 2016, 3c3.A viruses dominated towards the end of the North American season.
        However, these viruses did not successfully spread outside of North America.

        {% embed_image /reports/figures/feb-2018/h3n2_tree_lbi.png 12 %}

        <p>Fitness correlates of epitope mutations (<a href="https://biologydirect.biomedcentral.com/articles/10.1186/1745-6150-1-34">Wolf et al 2006</a>) and antigenic phenotype estimated from HI data (<a href="http://bedford.io/papers/neher-titers/">Neher et al 2016</a>) suggest that intrinsic fitness of clade A1b viruses may be slightly greater than clade A2 viruses.
        Using these correlates in a formal fitness model, as originally described by <a href="https://www.nature.com/articles/nature13087">&Lstrok;uksza and LÃ¤ssig</a>, suggests that clade A1b viruses will increase in frequency over the coming year and clade A2 viruses will decrease.
        However, fitness differences between these two clades are small and the model predicts that both will still circulate in Feb 2019.
        Additionally, this fitness model cannot capture the effects of the reassortment event and so it may not be properly weighting clade A2 viruses.

        {% embed_image /reports/figures/feb-2018/h3n2_freq_prediction.png 12 %}

        <p><i>We think that clade A1b viruses or reassortant clade A2 viruses are most likely to increase in the coming year.
        It is difficult to project a winner and cocirculation is a likely outcome.</i>

    </div>
    <div class="col-md-3">
        <div class="io-container title">
            A/H1N1pdm
        </div>
        <div class="spacer"></div>
    </div>
    <div class="col-md-9">

        <p><b>A clade comprising mutations S74R, S164T and I295V has recently swept to fixation.
        The rapidity of this sweep suggests a selective origin.
        However, there is no evidence of antigenic change.</b></p>

        <div class="spacer"></div>

        <p>We base our primary analysis on a set of viruses collected between Mar 2016 and Jan 2018, comprising approximately 100 viruses per month during relevant months of 2017.
        We use all available data when estimating frequencies of mutations and weight samples appropriately by regional population size and relative sampling intensity to arrive at a putatively unbiased global frequency estimate.
        Phylogenetic analyses are based on a representative sample of about 2000 viruses.

        {% embed_image /reports/figures/feb-2018/h1n1pdm_counts.png 12 %}

        <p>H1N1pdm recently underwent a selective sweep of the 6b.1 clade comprising mutations S162N and I216T.
        All currently circulating H1N1pdm viruses are clade 6b.1.
        The H1N1pdm vaccine strain was updated in Sep 2016 to A/Michigan/45/2015 to match 6b.1 viruses.

        <p>Within 6b.1 there has emerged a subclade comprising S74R, S164T and I295V that has recently swept to fixation.

        {% embed_image /reports/figures/feb-2018/h1n1pdm_tree_clades.png 12 %}

        <p>Subsequently, several amino acid substitutions have emerged within the S164T clade.
        The S183P substitution was circulating outside of the S164T clade previously and now reemerged within the S164T clade, predominantly in North America.
        Furthermore, the T120A mutations has risen to about 40% frequency over the last few months.

        {% embed_image /reports/figures/feb-2018/h1n1pdm_freq_mutations.png 12 %}

        <p>In a pattern familiar in the recent history of H1N1pdm evolution, none of these mutations has obvious an obvious impact in antigenic essays, yet they exhibit rapid shifts in frequency and recurrent mutations are successful on different genetic backgrounds, suggesting potential selective origins.

        {% embed_image /reports/figures/feb-2018/h1n1pdm_tree_titers.png 12 %}

    </div>

    <div class="col-md-3">
        <div class="io-container title">
            B/Vic
        </div>
        <div class="spacer"></div>
    </div>
    <div class="col-md-9">

        <p><b>Very little B/Vic activity has been observed in recent months.
        A clade with a two codon deletion at sites HA1:162/163 has gradually risen in frequency.
        HI measurements suggest an 8 to 16-fold titer drop relative to the vaccine strain, but this antigenic change has not yet resulted in a rapid rise of this variant.</b></p>

        <div class="spacer"></div>

        <p>We base our primary analysis on a set of viruses collected between Mar 2016 and Jan 2018, comprising >150 viruses per month in Dec 2016 to Apr 2017 and 20-50 viruses per month more recently.
        We use all available data when estimating frequencies of mutations and weight samples appropriately by regional population size and relative sampling intensity to arrive at a putatively unbiased global frequency estimate.
        Phylogenetic analyses are based on a representative sample of about 2000 viruses.

        {% embed_image /reports/figures/feb-2018/vic_counts.png 12 %}

        <p>Recent B/Vic sequences fall into three noteworthy clades.
        Most importantly, a clade defined by a deletion of codons 162/163 along with mutations D129G and I180V emerged in 2016 and has been slowly spreading since.

        {% embed_image /reports/figures/feb-2018/vic_tree_clades.png 12 %}

        It is now at ~17% global frequency and appears at high frequency in recent viruses from Europe and North America, although these estimates are based on little recent data.
        Additionally, a clade defined by a I175V substitution in HA1 has slowly risen to moderate frequency in North America.
        Another clade defined by mutation K209N has been circulating at low frequency in Asia throughout last year.

        {% embed_image /reports/figures/feb-2018/vic_freq_mutations.png 12 %}

        <p>The deletion variant (DV) shows a strong antigenic impact with an 8 to 16-fold titer drop relative to B/Texas/2/2013.
        This would suggest increased fitness, however, the slow spread of the variant suggests that it may have instrinsic deficiencies in viral fitness.
        Still, if current trends continue, then the deletion variant is expected to slowly increase in frequency over the coming year.

        {% embed_image /reports/figures/feb-2018/vic_tree_titers.png 12 %}

        <p>An extremely simple logistic growth model where growth rate is calculated from increase in frequency over the past 12 months, places frequency of the deletion variant at ~50% in Jan 2019.
        We caution again over-interpretation of this model; our expectation is that the deletion variant will circulate at moderate frequency in the coming year.
        Overall, there was very little B/Vic activity resulting in few sequenced viruses; this reduces confidence in these patterns.

        {% embed_image /reports/figures/feb-2018/vic_growth.png 12 %}

    </div>
    <div class="col-md-3">
        <div class="io-container title">
            B/Yam
        </div>
        <div class="spacer"></div>
    </div>
    <div class="col-md-9">

      <p><b>Europe experienced a strong and early B/Yam season in absence of amino acid variation in HA or antigenic diversity.
      However, several mutations in NA have rapidly swept or risen to intermediate frequencies.</b>

      <div class="spacer"></div>

      <p>We base our primary analysis on a set of viruses collected between Mar 2016 and Jan 2018, comprising between 50 and 150 viruses per month during recent months of 2017 and 2018.
      The majority of recent B/Yam sequences are from Europe.
      We use all available data when estimating frequencies of mutations and weight samples appropriately by regional population size and relative sampling intensity to arrive at a putatively unbiased global frequency estimate.
      Phylogenetic analyses are based on a representative sample of about 2000 viruses.

      {% embed_image /reports/figures/feb-2018/yam_counts.png 12 %}

      <p>We observe very little variation among the HA segments of B/Yam viruses.
      Previously circulating variants have either gone to fixation (M251V) or gone extinct (K211R).
      This lack of amino acid variation is consistent with little antigenic variation in HI assays.

      {% embed_image /reports/figures/feb-2018/yam_tree_clades.png 12 %}

      <p>While little amino acid variation is observed in recent HA sequences, a number of substitutions in NA have risen to high frequency.

      {% embed_image /reports/figures/feb-2018/yam_na_tree_lbi.png 12 %}

      <p>The mutation 342K rose from low frequency a year ago close to fixation.
      Within this clade, the mutation S402P is globally at about 50% frequency, dominating Europe.
      Outside of the clade defined by 402P, two mutations rose to 10-20% since the summer:
      S246T is common in Asia, while A395S is more common in North America.
      At sites 395 and 342, we observe several independent mutations with significant circulation during the last 3 years.
      The significance of these mutations is unclear to us, but their rapid rise, recurrent mutation, and the unusual early dominance of B/Yam in Europe suggests a selective origin.

      {% embed_image /reports/figures/feb-2018/yam_na_freq_mutations.png 12 %}

      <p>The LBI, when applied to the NA tree of B/Yam, suggests that the clade defined by the N342K mutation will take over (it is already above 80%) and that within this clade the subclade defined by the R65H will dominate.
      However, the dominance of isolates from European might skew this inference.

      <hr>

      <p><i>We thank the Influenza Division at the US Centers for Disease Control and Prevention, the Victorian Infectious Diseases Reference Laboratory at the Australian Peter Doherty Institute for Infection and Immunity, the Influenza Virus Research Center at the Japan National Institute of Infectious Diseases, the Crick Worldwide Influenza Centre at the UK Francis Crick Institute for data sharing and feedback.
      We thank Barney Potter for data cleaning and organization and we thank John Huddleston for fitness model implementation.
      We thank David Wentworth, Rebecca Garten, Xiyan Xu and Jackie Katz for insight regarding analysis directions.</i>

    </div>


</div>
